README.md
# dirichlet (Dirichlet Distributions)
Contains the r,d functions for a dirichlet and generalized dirichlet distribution

[![Travis build status](https://travis-ci.org/bertcarnell/dirichlet.svg?branch=master)](https://travis-ci.org/bertcarnell/dirichlet)
[![Coverage status](https://codecov.io/gh/bertcarnell/dirichlet/branch/master/graph/badge.svg)](https://codecov.io/github/bertcarnell/dirichlet?branch=master)


ChangeLog
Version 1.0
3/15/2013
1. Initial release to R-Forge

Version 1.1
2018
1. Switch to github


tests/testthat.R
library(testthat)
library(dirichlet)

test_check("dirichlet")


tests/testthat/test-dirichlet.r
context("test-dirichlet")

describe("ddirichlet", {
  it("A two parameter Dirichlet is equivalent to a beta", {
    expect_equal(ddirichlet(c(0.3, 0.7), c(2, 3)), dbeta(0.3, 2, 3))
  })
  it("Dirichlet which is uniform across the parameter space has the appropriate density", {
    # uniform on 0,1 with height 2
    expect_equal(ddirichlet(c(.1,.8,.1), c(1,1,1)), 2)
    expect_equal(ddirichlet(c(.1,.6,.3), c(1,1,1)), 2)
  })
  it("That the function returns values equal to an independent calculation", {
    expect_equal(ddirichlet(c(.1,.4,.5), c(1,2,3)),
                       gamma(1 + 2 + 3)/gamma(1)/gamma(2)/gamma(3) *
                         .1^(1 - 1)*.4^(2 - 1)*.5^(3 - 1))
    expect_equal(ddirichlet(c(.1,.2,.7), c(4,5,8)),
                       gamma(4 + 5 + 8)/gamma(4)/gamma(5)/gamma(8) *
                         .1^(4 - 1)*.2^(5 - 1)*.7^(8 - 1))
  })
  it("That values of a outside the Dirichlet parameters yield values of zero", {
    expect_equal(ddirichlet(c(1, 1, 1), c(1,1,1)), 0)
    expect_equal(ddirichlet(c(.1, .1, .1), c(1,1,1)), 0)
    expect_equal(ddirichlet(c(-1, .1, .1), c(4, 3, 2)), 0)
    expect_equal(ddirichlet(c(.1, 2, .1), c(4, 3, 2)), 0)
  })
})

describe("rdirichlet", {
  set.seed(1976)
  n <- 10000
  p <- c(.4, .3, .2, .1)
  Y <- rdirichlet(n, p)
  expect_true(all(dim(Y) == c(n, length(p))))

  alpha <- c(4,8,2)
  Y2 <- rdirichlet(n, alpha)
  expect_true(all(dim(Y2) == c(n, length(alpha))))

  Y3 <- rdirichlet(n, p, allowZero = TRUE)
  expect_true(all(dim(Y3) == c(n, length(p))))
  it("The returned vector is on (0,1)",{
    expect_true(all(Y >= 0 && Y <= 1))
    expect_true(all(Y2 >= 0 && Y2 <= 1))
    expect_true(all(Y3 >= 0 && Y3 <= 1))
  })
  it("The sum of the values in each draw is one.", {
    expect_equal(rowSums(Y), rep(1, n))
    expect_equal(rowSums(Y2), rep(1, n))
    expect_equal(rowSums(Y3), rep(1, n))
  })
  it("The mean of all the draws for each parameter is equal to the desired mean probabilities", {
    expect_equal(apply(Y, 2, mean), p, tol = 0.05)
    expect_equal(apply(Y2, 2, mean), alpha/sum(alpha), tol = 0.05)
    expect_equal(apply(Y3, 2, mean), p, tol = 0.05)
  })
  expect_error(Y4 <- rdirichlet(100, c(0.8, 0, 0.2)))
  Y4 <- rdirichlet(100, c(0.8, 0, 0.2), allowZero = TRUE)
  expect_equal(Y4[,2], rep(0, 100))
})

describe("qdirichlet",{
  set.seed(1976)
  X <- matrix((1:12)/13, ncol = 3, nrow = 4)
  X2 <- X
  X2[2,3] <- NA
  A <- matrix(
    c(0.01605770, 0.2673003, 0.7166420,
      0.02892119, 0.2698036, 0.7012752,
      0.03886065, 0.2672579, 0.6938815,
      0.04514385, 0.2554320, 0.6994241), nrow = 4, ncol = 3, byrow = TRUE)

  B <- matrix(
    c(0.02191580, 0, 0.9780842,
      0.03960742, 0, 0.9603926,
      0.05303454, 0, 0.9469655,
      0.06063094, 0, 0.9393691), nrow = 4, ncol = 3, byrow = TRUE)
  assert_that(all(abs(rowSums(A) - rep(1,4)) < 1E-7))
  assert_that(all(abs(rowSums(B) - rep(1,4)) < 1E-7))
  it("The result of the function matches known output within 1E-7", {
    expect_equal(qdirichlet(X, c(1,2,3)), A, tolerance = 1E-7)
    expect_equal(qdirichlet(X, c(1,0,3)), B, tolerance = 1E-7)
  })
  it("Errors are generated for illegal input", {
    expect_error(qdirichlet(c(2,3), c(1,2,3)), silent = TRUE)
    expect_error(qdirichlet(matrix(nrow = 3, ncol = 2), c(1,2,3)), silent = TRUE)
    expect_error(qdirichlet(X2, c(1,2,3)), silent = TRUE)
    expect_error(qdirichlet(X, c(1,NA,3)), silent = TRUE)
  })
  it("Other", {
    Z <- qdirichlet(rdirichlet(100, c(3,4,5)), c(3,4,5))
    expect_true(all( Z <= 1 & Z >= 0))
    expect_equal(rowSums(Z), rep(1, 100))

    # test for numerical instability in the qgamma calculation
    expect_warning(Z <- qdirichlet(matrix(c(.001, .1, .899,
                             .001, .1, .899), nrow = 2, ncol = 3, byrow = TRUE),
                    c(.001, 2, 2)))
    expect_false(any(is.nan(Z)))

    A <- matrix(runif(20*1000), nrow = 1000, ncol = 20)
    A <- A / rep(rowSums(A), 20)
    expect_warning(Z <- qdirichlet(A, c(.001, .01, .1, 1, rep(2, 16))))
    expect_false(any(is.nan(Z)))

    expect_warning(qdirichlet(matrix(c(.001, .1, .899,
                                       .001, .1, .899), nrow = 2, ncol = 3, byrow = TRUE),
                              c(.001, 2, 2)), silent = TRUE)
  })
})

test_that("fit.dirichlet", {
  tempTest <- function(desired_p, desired_k, type="mm")
  {
    set.seed(1976)
    Y <- rdirichlet(1000, desired_k*desired_p)
    fit.dirichlet(Y, type)
  }

  desired_p <- c(.4, .3, .2, .1)
  temp <- tempTest(desired_p, 5)
  expect_equal(temp$most.likely.k, 5, tolerance = 0.5)
  expect_equal(temp$weighted.k, 5, tolerance = 0.5)
  expect_equal(temp$p, desired_p, tolerance = 0.05)

  desired_p <- c(.5, .3, .2)
  temp <- tempTest(desired_p, 100)
  expect_equal(temp$most.likely.k, 100, tolerance = 0.5)
  expect_equal(temp$weighted.k, 100, tolerance = 0.5)
  expect_equal(temp$p, desired_p, tolerance = 0.05)

  desired_p <- c(.4, .3, .2, .1)
  temp <- tempTest(desired_p, 5, "ml")
  expect_equal(temp$k, 5, tolerance = 0.5)
  expect_equal(temp$p, desired_p, tolerance = 0.05)

  desired_p <- c(.5, .3, .2)
  temp <- tempTest(desired_p, 100, "ml")
  expect_equal(temp$k, 100, tolerance = .1)
  expect_equal(temp$p, desired_p, tolerance = 0.05)

  expect_error(fit.dirichlet(4), silent = TRUE)
  expect_error(fit.dirichlet(matrix(c(1,2,3,4), nrow = 2), "not there yet"))
})

test_that("qMarginalDirichlet", {
  expect_equal(rep(0.99, 3), qMarginalDirichlet(c(0.9999), c(2, 2, 2)), tolerance = 0.01)
  expect_error(qMarginalDirichlet(c(1, 0.2, 0.1), c(4, 3, 2)))
  expect_error(qMarginalDirichlet(c(0.7, 0.2, 0.1), c(4, -1, 2)))
})

test_that("calculateDirichletCV", {
  expect_equal(rep(sqrt(2)/2, 3), calculateDirichletCV(c(1,1,1)), tolerance = 1E-6)
})


tests/testthat/test-gendirichlet.r
context("test-gendirichlet")

test_that("dGenDirichlet", {
  expect_equal(0, dGenDirichlet(c(0.1, 0.1, 0.1), c(0.7, 0.2, 0.1), c(2, 3, 4)))
  expect_equal(0, dGenDirichlet(c(0.8, 0.1, 0.1), c(1.1, 0.2, 0.1), c(2, 3, 4)))
  expect_equal(0, dGenDirichlet(c(0.8, 0.1, 0.1), c(0.7, 0.2, -0.1), c(2, 3, 4)))
  expect_equal(0, dGenDirichlet(c(0.8, 0.1, 0.1), c(0.7, 0.2, 0.1), c(2, -3, 4)))
})

test_that("dGenDirichletStd", {
  expect_equal(0, dGenDirichletStd(c(0.1, 0.1, 0.1), c(0.7, 0.2, 0.1), c(2, 3, 4)))
})

describe("rGenDirichlet", {
  set.seed(1976)
  n <- 10000
  p <- c(.4, .3, .2, .1)
  k <- c(10, 20, 30, 40)
  Y <- rGenDirichlet(n, p, k)
  it("The returned vector is on (0,1)", {
    expect_true(all(dim(Y) == c(n, length(p))))
    expect_true(all(Y >= 0 && Y <= 1))
  })
  it("The sum of the values in each draw is one.", {
    expect_equal(rowSums(Y), rep(1, n))
  })
  it("The mean of all the draws for each parameter is equal to the desired mean probabilities", {
    expect_equal(apply(Y, 2, mean), p, tol = 0.05)
  })
  it("Errors are generated on illegal inputs", {
    p <- c(.1, .7, .05, .15)
    expect_error(rGenDirichlet(-n, p, k))
    expect_warning(rGenDirichlet(10, c(.5, .5, .5), c(2, 3, 4)))
  })
  it("That rdirichlet and rGenDirichlet are equivalent", {
    Y <- matrix(c(.5, .3, .2), nrow = 1, ncol = 3)
    p <- c(.4, .35, .25)
    A <- rGenDirichlet(n, p, c(2,1.2,0))
    B <- rdirichlet(n, p*2)
    expect_equal(apply(A,2,var), apply(B,2,var), tolerance = 0.01)
  })
})

describe("qGenDirichlet", {
  set.seed(1976)
  n <- 100
  p <- c(.4, .3, .2, .1)
  k <- c(10, 20, 30, 40)
  Y <- rGenDirichlet(n, p, k)
  Z <- qGenDirichlet(Y, p, k)
  Z2 <- qGenDirichlet(Y, c(.4, 0, .35, .25), k)
  Z3 <- qGenDirichlet(Y, c(.4, .35, 0, .25), c(2, 3, 4, 5))
  W <- qGenDirichlet(Y, c(.35, .4, 0, .25), c(3, 2, 4, 5))
  it("The quantiles are on (0,1)", {
    expect_true(all( Z <= 1 & Z >= 0))
    expect_equal(rowSums(Z), rep(1, n))
    expect_true(all(Z2 <= 1 & Z2 >= 0))
    expect_equal(rowSums(Z2), rep(1, n))
    expect_true(all(Z2[,2] == 0))
  })
  it("Errors are generated for illegal input", {
    expect_error(rGenDirichlet(-1, p, k))
    expect_warning(qGenDirichlet(Y, c(0.5, 0.5, 0.5, 0.5), c(2, 3, 4, 5)))
  })
  expect_equal(apply(Z3, 2, mean), apply(W, 2, mean)[c(2,1,3,4)])
})

test_that("fit_genDirichlet", {
  tempTest <- function(desired_p, desired_k, type="mm", N)
  {
    set.seed(1976)
    Y <- rGenDirichlet(N, desired_p, desired_k)
    fit.genDirichlet(Y, type)
  }

  desired_p <- c(.4, .3, .2, .1)
  desired_k <- c(5, 6, 7, 10)
  temp <- tempTest(desired_p, desired_k, N = 10000)
  expect_equal(temp$k[1:3], desired_k[1:3], tolerance = 0.05*max(desired_k))
  expect_equal(temp$p, desired_p, tolerance = 0.01)

  desired_p <- c(.5, .3, .2)
  desired_k <- c(50, 60, 70)
  temp <- tempTest(desired_p, desired_k, N = 10000)
  expect_equal(temp$k[1:2], desired_k[1:2], tolerance = 0.05*max(desired_k))
  expect_equal(temp$p, desired_p, tolerance = 0.01)

  # this test failed before adding the sumx numerical accuracy code
  a <- c(1, 10, 20, 40, 100, 500)
  p1 <- 1/a / sum(1/a)
  a <- c(1, 10, 10, 50, 200, 200)
  p2 <- 1/a / sum(1/a)
  set.seed(1876)
  Y <- rdirichlet(100, p1*3)
  Z <- rdirichlet(100, p2*4)
  X <- as.matrix(rbind(Y, Z))
  temp <- fit.genDirichlet(X)
  expect_true(all(is.finite(temp$k)))
  expect_true(all(is.finite(temp$p)))

  desired_p <- c(.4, .3, .2, .1)
  desired_k <- c(5, 6, 7, 10)
  temp <- tempTest(desired_p, desired_k, "mm", N = 10000)
  expect_equal(temp$k[1:3], desired_k[1:3], tolerance = 0.05*max(desired_k))
  expect_equal(temp$p, desired_p, tolerance = 0.1)

  desired_p <- c(.5, .3, .2)
  desired_k <- c(50, 60, 70)
  temp <- tempTest(desired_p, desired_k, "mm", N = 10000)
  expect_equal(temp$k[1:2], desired_k[1:2], tolerance = 0.05*max(desired_k))
  expect_equal(temp$p, desired_p, tolerance = 0.1)

  expect_error(fit.dirichlet(4))
  expect_error(fit.dirichlet(matrix(c(1,2,3,4), nrow = 2), "not there yet"))

  desired_p <- c(.5, .3, .2)
  desired_k <- c(5, 6, 7)
  temp <- tempTest(desired_p, desired_k, "ml", N = 100)
  expect_equal(temp$k[1:2], desired_k[1:2], tolerance = 0.05*max(desired_k))
  expect_equal(temp$p, desired_p, tolerance = 0.05)
})

test_that("qApproxMarginalGenDirichlet", {
  X <- qApproxMarginalGenDirichlet(c(0.05,0.5,0.95), c(.4,.3,.2,.1), c(3,4,5,6))
  expect_equal(dim(X)[1], 3)
  expect_equal(dim(X)[2], 4)
  expect_true(all(X < 1 & X > 0))
  expect_equal(X[2,1], qbeta(0.5, 3*.4, 3*.6))
  expect_error(qApproxMarginalGenDirichlet(.5, .5, 2))
  expect_error(qApproxMarginalGenDirichlet(.5, c(.6,.4), c(3,4,5)))
  expect_error(qApproxMarginalGenDirichlet(.5, c(.6,.4), c(3,4)))
  expect_error(qApproxMarginalGenDirichlet(.5, c(.2,.3,.5), c(3,4,5)))
  expect_warning(qApproxMarginalGenDirichlet(c(0.05,0.5,0.95), c(.5,.5,.5,.5), c(3,4,5,6)))
})

test_that("calculateGenDirichletCV", {
  expect_equal(1, calculateGenDirichletCV(c(0.5, 0.3, 0.2), c(1, 1, 1))[2])
})

test_that("calculateConstantCVGenDirichletK", {
  expect_warning(k <- calculateConstantCVGenDirichletK(c(.5,.3,.2), 10))
  expect_true(length(k) == 3)
  expect_true(all(k[1:2] > 0))
  expect_equal(k[1], 10)

  a <- c(1, 10, 20, 40, 100, 500)
  p1 <- 1/a / sum(1/a)
  expect_warning(k <- calculateConstantCVGenDirichletK(p1, 3))
  expect_true(length(k) == 6)
  expect_true(all(k[1:5] > 0))
  expect_equal(k[1], 3)

  expect_warning(k <- calculateConstantCVGenDirichletK(c(.4,.3,.2,.1), 1))
  expect_equal(k, c(1, 2.2, .Machine$double.xmax, .Machine$double.xmax))

  k <- calculateConstantCVGenDirichletK(c(.3,.2,.15,.10,.05, .05, .05, .05, .05), 20)
  expect_equal(k[c(1,9)], c(20,  32.02777777777888))

  expect_warning(calculateConstantCVGenDirichletK(c(.5,.5,.5), 2))
})


R/dirichlet.r
# Copyright 2018 Robert Carnell


#' Dirichlet Functions
#'
#' @description  pdf and random deviates of the Dirichlet distribution.
#'
#' @param x vector of quantiles.  For \code{x < 0 || x > 1} or if the sum of \code{x} is not equal to 1, zero will be returned for the density
#' @param alpha The Dirichlet parameters.  \code{alpha > 0} unless \code{allowZero=TRUE}
#' @param n number of observations
#' @param allowZero Indicator that zeros are allowed in the alpha vector and should be handled appropriately
#'
#' @return ddirichlet gives the density and rdirichlet generates random deviates.
#' @name dirichlet
#' @importFrom assertthat assert_that
#' @export
#'
#' @examples
#' rdirichlet(10, c(4,3,2))
#' ddirichlet(c(.2, .7, .1), c(4, 3, 2))
#' rdirichlet(5, c(4,3,0,2), allowZero=TRUE)
#' (ddirichlet(c(-0.2, 1.2), c(2,2)) == 0)
ddirichlet <- function(x, alpha) {
  ## probability density for the Dirichlet function
  ## x = vector of values at which the density is desired
  ## alpha = vector of dirichlet parameters ( k * p )

  assert_that(length(alpha) == length(x), msg = "the length of x and alpha must be equal")
  assert_that(all(alpha > 0), msg = "All alphas must be > 0")
  # the density outside the support space for x is zero
  if (any(x <= 0))
    return(0)
  if (any(x > 1))
    return(0)
  if (abs(sum(x) - 1) > 10*.Machine$double.eps)
    return(0)

  s <- sum((alpha - 1 ) * log(x))
  return(exp(s - (sum(lgamma(alpha)) - lgamma(sum(alpha)))))
}

################################################################################

#' @rdname dirichlet
#' @export
rdirichlet <- function(n, alpha, allowZero=FALSE) {
  ## pick n random deviates from the Dirichlet function with shape
  ## parameters alpha.  alpha = 0 is allowed and produces 0 for those options

  lena <- length(alpha)
  assert_that(lena > 1 && n > 0, msg = "The length of alpha must be > 1 and n > 0")
  if (!allowZero)
  {
    assert_that(all(alpha > 0), msg = "All alpha must be >0 when allowZer0 = FALSE")
    x <- matrix( rgamma(lena*n, alpha), ncol = lena, byrow = TRUE)
    sm <- x %*% rep(1, lena)
    return(x / as.vector(sm))
  } else
  {
    assert_that(all(alpha >= 0), msg = "All alpha must be >= 0")
    ind <- which(alpha != 0)
    X <- sapply(alpha[ind], function(x) rgamma(n, x, 1))
    Xsum <- apply(X, 1, sum)
    Y <- apply(X, 2, "/", Xsum)
    Z <- matrix(0, nrow = n, ncol = length(alpha))
    Z[,ind] <- Y
    return(Z)
  }
}

################################################################################


#' Approxminate Dirichlet Quantiles
#'
#' qdirichlet is not an exact quantile function since the quantile of a
#' multivariate distribtion is not unique.
#' qdirichlet is also not the quantiles of the marginal distributions since
#' those quantiles do not sum to one
#' qdirichlet is the quantile of the underlying gamma functions, normalized
#' This has been tested to show that qdirichlet approximates the dirichlet
#' distribution well and creates the correct marginal means and variances
#' when using a latin hypercube sample
#' @param X a matrix of marginal probabilities
#' @param alpha The Dirichlet parameters.  \code{alpha > 0}
#'
#' @return approximate dirichlet quantiles
#' @importFrom stats pgamma qgamma rgamma
#' @export
#'
#' @examples
#' qdirichlet(matrix(c(0.1, 0.1, 0.3), nrow=1, byrow=TRUE), c(2,3,4))
qdirichlet <- function(X, alpha) {
  lena <- length(alpha)
  assert_that(is.matrix(X), msg = "X must be a matrix")
  sims <- dim(X)[1]
  assert_that(dim(X)[2] == lena, msg = "The number of columns of X must equal the length of alpha")
  if (any(is.na(alpha)) || any(is.na(X)))
    stop("NA values not allowed in qdirichlet")

  Y <- matrix(0, nrow = sims, ncol = lena)
  ind <- which(alpha != 0)
  for (i in ind)
  {
    # add check to trap numerical instability in qgamma
    # start to worry if alpha is less than 1.0
    if (alpha[i] < 1)
    {
      # look for places where NaN will be returned by qgamma
      nanind <- which(pgamma(.Machine$double.xmin, alpha[i], 1) >= X[,i])
      # if there are such places
      if (length(nanind) > 0)
      {
        # set the output probability to near zero
        Y[nanind,i] <- .Machine$double.xmin
        # calculate the rest
        Y[-nanind,i] <- qgamma(X[-nanind,i], alpha[i], 1)
        warning("at least one probability set to the minimum machine double")
      } else {
        Y[,i] <- qgamma(X[,i], alpha[i], 1)
      }
    } else {
      Y[,i] <- qgamma(X[,i], alpha[i], 1)
    }
  }
  Y <- Y / rowSums(Y)
  return(Y)
}

################################################################################

#' Fit Dirichlet parameters
#'
#' @param X samples from a dirichlet distribution
#' @param type \code{mm} for Method of Moments estimation or \code{ml} for Maximum Likelihood estimation.
#'
#' @return A list giving the \code{k} value or vector of \code{k} values for the Generalized
#' Dirichlet and vector of \code{p} estimates.  The normal Dirichlet parameterization can be obtained by \code{k*p}.
#' The \code{fit.dirichlet} function returns two estimates of \code{k}.  One based on the most likely
#' parameter in the distribution, the other based on a weighted mean of the \code{k} estimates for each parameter.
#'
#' @importFrom stats weighted.mean optim var
#' @name fit_dirichlet
#' @export
#'
#' @examples
#' fit.dirichlet(matrix(c(.1, .9, .2, .8, .3, .7), nrow=3, ncol=2, byrow=TRUE), type="mm")
#' Z <- rdirichlet(100, c(4,3,2))
#' fit.dirichlet(Z)
#' fit.dirichlet(Z, "ml")
fit.dirichlet <- function(X, type="mm")
{
  assert_that(is.matrix(X), msg = "X must be a matrix")

  # get method of moments estimates to return to use as starting values for the
  #  maximum likelihood method
  p <- apply(X, 2, mean)
  ind <- which.max(p)
  v <- apply(X, 2, var)
  temp <- p*(1 - p)/v - 1
  #alpha <- p[ind]*temp
  #beta <- (1-p[ind])*temp
  #k <- alpha + beta

  if (type == "mm")
  {
    return(list(most.likely.k = temp[ind], weighted.k = weighted.mean(temp, v), p = p))
  } else if (type == "ml")
  {
    f <- function(a)
    {
      # pass in a on the log scale so that it is guaranteed to be a strictly
      #  positive number when transformed back
      a <- exp(a)
      return(-sum(log(apply(X, 1, ddirichlet, alpha = a))))
    }

    #  use starting values on the log scale
    o <- optim(log(temp[ind]*p), f)

    if (o$convergence != 0)
      stop(o$message)

    # the probabilities are the normalized a values
    p <- exp(o$par)
    p <- p / sum(p)

    k <- exp(o$par)[ind]/p[1]

    return(list(k = k, p = p))
  } else stop("type not recognized")
}

################################################################################

#' Quantiles of the marginal beta distributions in a dirichlet distribution
#'
#' @param p the probability of the desired quantile
#' @param alpha the vector of dirichlet parameters
#'
#' @return the vector of marginal quantiles
#' @importFrom stats qbeta
#' @export
#'
#' @examples
#' qMarginalDirichlet(c(0.1, 0.2, 0.4), c(2,3,4))
qMarginalDirichlet <- function(p, alpha) {
  # the marginals of the Dirichlet distribution are beta distributions
  #  with paramters alpha = k*p, beta = k(1-p)
  #  where the Dirichlet alpha = k*p
  #  p is the vector of desired percentiles
  # returns a length p x length alpha matrix

  assert_that(all(p > 0 & p < 1), msg = "p must be > 0 and < 1")
  assert_that(all(alpha > 0), msg = "all alpha must be > 0")

  p_dirichlet <- alpha / sum(alpha)
  k <- alpha[1] / p[1]

  sapply(p_dirichlet, function(x) qbeta(p, k*x, k*(1 - x)))
}

################################################################################

#' The coefficient of variation of the marginal betas
#'
#' @param alpha the dirichlet parameters
#'
#' @return the marginal CVs
#' @export
#'
#' @examples
#' calculateDirichletCV(c(5, 4, 2))
calculateDirichletCV <- function(alpha) {
  p <- alpha / sum(alpha)
  k <- sum(alpha)
  margCV <- sqrt(p*(1 - p)/(k + 1))/p

  return(margCV)
}


trunk/RUnit/runitTestSuite_dirichlet.R
################################################################################
#
# Program Name:  runitTestSuite_dirichlet.R
# Purpose:       To provide test functions for the dirichlet package
# Author:        Rob Carnell
# Date:          July 2007
#
# Required Packages:  RUnit
# R version:          2.4.0 (>=2.3.1)
#
################################################################################

rm(list=ls())
require(RUnit)

## used as part of package
#require(dirichlet)
#defaultPath <- chartr("/", "//", paste(.path.package("dirichlet"), "/RUnit", sep=""))

################# used in development ##########################################
defaultPath <- file.path("c:", "Documents and Settings", "carnellr",
                         "My Documents", "Repositories", "BTRA",
                         "Trunk", "Source", "Rscripts", "dirichlet")
                         
temp <- list.files(file.path(defaultPath, "R"), full.names=TRUE)
junk <- lapply(temp, source)

testSuite.d <-
  defineTestSuite("dirichlet", dirs=file.path(defaultPath, "RUnit"),
                  testFileRegexp="^runit_[[:alnum:]]+.[rR]$")

testResult <- runTestSuite(testSuite.d)

################# used in development ##########################################

htmlFile <- file.path(defaultPath, "RUnit", "Test Results.html")

## warning expected about gcc compiler
suppressWarnings(printHTMLProtocol(testResult, fileName=htmlFile))

browseURL(htmlFile, browser=getOption("browser"))



trunk/RUnit/runit_testdirichlet.R
################################################################################
# Program Name:  runit_dirichlet.R
# Purpose:       Test dirichlet functions
# Charge:        G882602
# Author:        Rob Carnell
# Date:          December 2006
#
# R version:     >= 2.4.1
#
################################################################################

#/**
# /testedfunction ddirichlet
# /testedpurpose  The density of the Dirichlet distribution
# /test           1. That a two parameter Dirichlet is equivalent to a beta
# /test           2. That a Dirichlet which is uniform across the parameter space has the appropriate density
# /test           3. That the function returns values equal to an independent calculation
# /test           4. That values of a outside the Dirichlet parameters yield values of zero
#*/
test.ddirichlet <- function()
{
  # generalization of the beta
  checkEqualsNumeric(ddirichlet(c(0.3, 0.7), c(2, 3)), dbeta(0.3, 2, 3))
  # uniform on 0,1 with height 2
  checkEqualsNumeric(ddirichlet(c(.1,.8,.1), c(1,1,1)), 2)
  checkEqualsNumeric(ddirichlet(c(.1,.6,.3), c(1,1,1)), 2)
  # non-uniform
  checkEqualsNumeric(ddirichlet(c(.1,.4,.5), c(1,2,3)),
                     gamma(1+2+3)/gamma(1)/gamma(2)/gamma(3)*
                     .1^(1-1)*.4^(2-1)*.5^(3-1))
  checkEqualsNumeric(ddirichlet(c(.1,.2,.7), c(4,5,8)),
                     gamma(4+5+8)/gamma(4)/gamma(5)/gamma(8)*
                     .1^(4-1)*.2^(5-1)*.7^(8-1))
  checkEqualsNumeric(ddirichlet(c(1,1,1), c(1,1,1)), 0)
  checkEqualsNumeric(ddirichlet(c(.1,.1,.1), c(1,1,1)), 0)
}

#/**
# /testedfunction rdirichlet
# /testedpurpose  a random draw from the multi-variate Dirichlet distribution
# /test           1. The returned vector is on (0,1)
# /test           2. The sum of the values in each draw is one.
# /test           3. The mean of all the draws for each parameter is equal to the desired mean probabilities
#*/
test.rdirichlet <- function()
{
  set.seed(1976)
  n <- 100000
  p <- c(.4, .3, .2, .1)
  Y <- rdirichlet(n, p)
  checkTrue(all(dim(Y) == c(n, length(p))))
  checkTrue(all(Y >= 0 && Y <= 1))
  checkEqualsNumeric(rowSums(Y), rep(1, n))
  checkEqualsNumeric(apply(Y, 2, mean), p, tol=.05)

  alpha <- c(4,8,2)
  Y <- rdirichlet(n, alpha)
  checkTrue(all(dim(Y) == c(n, length(alpha))))
  checkTrue(all(Y >= 0 && Y <= 1))
  checkEqualsNumeric(rowSums(Y), rep(1, n))
  checkEqualsNumeric(apply(Y, 2, mean), alpha/sum(alpha), tol=.05)
}

#/**
# /testedfunction qdirichlet
# /testedpurpose  Find the marginal quantiles of a dirichlet distribution
# /test           1. The result of the function matches known output within 1E-7
# /test           2. Errors are generated for illegal input
#*/
test.qdirichlet <- function()
{
  set.seed(1976)
  X <- matrix((1:12)/13, ncol=3, nrow=4)
  X2 <- X; X2[2,3] <- NA
  A <- matrix(
       c(0.01605770, 0.2673003, 0.7166420,
         0.02892119, 0.2698036, 0.7012752,
         0.03886065, 0.2672579, 0.6938815,
         0.04514385, 0.2554320, 0.6994241), nrow=4, ncol=3, byrow=TRUE)

  B <- matrix(
       c(0.02191580, 0, 0.9780842,
         0.03960742, 0, 0.9603926,
         0.05303454, 0, 0.9469655,
         0.06063094, 0, 0.9393691), nrow=4, ncol=3, byrow=TRUE)
         
  checkEqualsNumeric(rowSums(A), rep(1,4), tolerance=1E-7)
  checkEqualsNumeric(rowSums(B), rep(1,4), tolerance=1E-7)

  checkException(qdirichlet(c(2,3), c(1,2,3)), silent=TRUE)
  checkException(qdirichlet(matrix(nrow=3, ncol=2), c(1,2,3)), silent=TRUE)
  checkException(qdirichlet(X2, c(1,2,3)), silent=TRUE)
  checkException(qdirichlet(X, c(1,NA,3)), silent=TRUE)
  checkEqualsNumeric(qdirichlet(X, c(1,2,3)), A, tolerance=1E-7)
  checkEqualsNumeric(qdirichlet(X, c(1,0,3)), B, tolerance=1E-7)

  Z <- qdirichlet(rdirichlet(100, c(3,4,5)), c(3,4,5))
  checkTrue(all( Z <= 1 & Z >= 0))
  checkEqualsNumeric(rowSums(Z), rep(1, 100))

  # test for numerical instability in the qgamma calculation
  Z <- qdirichlet(matrix(c(.001, .1, .899,
                      .001, .1, .899), nrow=2, ncol=3, byrow=TRUE),
                      c(.001, 2, 2))
  checkTrue(any(is.nan(Z)) == FALSE)

  A <- matrix(runif(20*1000), nrow=1000, ncol=20)
  A <- A / rep(rowSums(A), 20)
  Z <- qdirichlet(A, c(.001, .01, .1, 1, rep(2, 16)))
  checkTrue(any(is.nan(Z)) == FALSE)
  
  options(warn=2)
  checkException(qdirichlet(matrix(c(.001, .1, .899,
                      .001, .1, .899), nrow=2, ncol=3, byrow=TRUE),
                      c(.001, 2, 2)), silent=TRUE)
  options(warn=1)

}

test.fit_Dirichlet <- function()
{
  tempTest <- function(desired_p, desired_k, type="mm")
  {
    set.seed(1976)
    Y <- rdirichlet(1000, desired_k*desired_p)
    fit.dirichlet(Y, type)
  }
  
  desired_p <- c(.4, .3, .2, .1)
  temp <- tempTest(desired_p, 5)
  checkEqualsNumeric(temp$most.likely.k, 5, tolerance=0.5)
  checkEqualsNumeric(temp$weighted.k, 5, tolerance=0.5)
  checkEqualsNumeric(temp$p, desired_p, tolerance=0.05)

  desired_p <- c(.5, .3, .2)
  temp <- tempTest(desired_p, 100)
  checkEqualsNumeric(temp$most.likely.k, 100, tolerance=0.5)
  checkEqualsNumeric(temp$weighted.k, 100, tolerance=0.5)
  checkEqualsNumeric(temp$p, desired_p, tolerance=0.05)

  desired_p <- c(.4, .3, .2, .1)
  temp <- tempTest(desired_p, 5, "ml")
  checkEqualsNumeric(temp$k, 5, tolerance=0.5)
  checkEqualsNumeric(temp$p, desired_p, tolerance=0.05)

  desired_p <- c(.5, .3, .2)
  temp <- tempTest(desired_p, 100, "ml")
  checkEqualsNumeric(temp$k, 100, tolerance=.1)
  checkEqualsNumeric(temp$p, desired_p, tolerance=0.05)

  checkException(fit.dirichlet(4), silent=TRUE)
  checkException(fit.dirichlet(matrix(1,2,3,4, nrow=2), "not there yet"), silent=TRUE)
}

